src/Common/FitOnFhir.Common/Services/UsersServiceBase.cs

```diff
@@ -0,0 +1,49 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using EnsureThat;
+using FitOnFhir.Common.Models;
+using FitOnFhir.Common.Repositories;
+using Hl7.Fhir.Model;
+using Microsoft.Health.Extensions.Fhir.Service;
+
+namespace FitOnFhir.Common.Services
+{
+    public abstract class UsersServiceBase
+    {
+        private readonly ResourceManagementService _resourceManagementService;
+        private readonly IUsersTableRepository _usersTableRepository;
+
+        public UsersServiceBase(ResourceManagementService resourceManagementService, IUsersTableRepository usersTableRepository)
+        {
+            _resourceManagementService = EnsureArg.IsNotNull(resourceManagementService, nameof(resourceManagementService));
+            _usersTableRepository = EnsureArg.IsNotNull(usersTableRepository, nameof(usersTableRepository));
+        }
+
+        public async System.Threading.Tasks.Task EnsurePatientAndUser(string platformName, string identifier, string system, CancellationToken cancellationToken)
+        {
+            EnsureArg.IsNotNullOrWhiteSpace(platformName, nameof(platformName));
+            EnsureArg.IsNotNullOrWhiteSpace(identifier, nameof(identifier));
+            EnsureArg.IsNotNullOrWhiteSpace(system, nameof(system));
+
+            // Get or create a Patient resource.
+            Patient patient = await _resourceManagementService.EnsureResourceByIdentityAsync<Patient>(
+                identifier,
+                system,
+                (p, id) => p.Identifier = new List<Identifier> { id });
+
+            // Check to see if the user exists in the repository.
+            User user = await _usersTableRepository.GetById(patient.Id, cancellationToken);
```

Comment:
Wondering what happens, the way this is written, if a user exists already but with a different platform saved, then won't this skip saving the google fit platform info below?

Reply:
Good catch. We should talk this through on Monday. Basically, we need to have an external identifier to resolve this problem. When a user starts the authorization flow, they might need to pass in an identifier (or a token oid once the method is no longer anonymous) that we can use to link all platforms to a single person.